<div style="text-align:left;">
  <a href="https://code213.tech/" target="_blank">
    <img src="../images/code213.PNG" alt="QWorld">
  </a>
  <p><em>prepared by Latreche Sara</em></p>
</div>

# Cluster the Olivetti Faces Dataset

## Load the Olivetti Faces Dataset

In [ ]:
# Step 1: Import the dataset loading function from scikit-learn
# Hint: Use the function `fetch_olivetti_faces()` from the module `sklearn.datasets`.

#  Step 2: Load the Olivetti faces dataset
# Hint: Call the function `fetch_olivetti_faces()` and assign it to a variable named `olivetti`.
# This function returns a dictionary-like object with face data, labels, images, and a description.


In [ ]:
from sklearn.datasets import fetch_olivetti_faces

olivetti= fetch_olivetti_faces()

downloading Olivetti faces from https://ndownloader.figshare.com/files/5976027 to /root/scikit_learn_data


In [ ]:
# 📄 Step 3: Display the dataset description
# Hint: The 'olivetti' object has an attribute called 'DESCR' that contains a description of the dataset.
# This will help you understand the background and structure of the data you're working with.


In [ ]:
print(olivetti.DESCR)

.. _olivetti_faces_dataset:

The Olivetti faces dataset
--------------------------

`This dataset contains a set of face images`_ taken between April 1992 and
April 1994 at AT&T Laboratories Cambridge. The
:func:`sklearn.datasets.fetch_olivetti_faces` function is the data
fetching / caching function that downloads the data
archive from AT&T.

.. _This dataset contains a set of face images: https://cam-orl.co.uk/facedatabase.html

As described on the original website:

    There are ten different images of each of 40 distinct subjects. For some
    subjects, the images were taken at different times, varying the lighting,
    facial expressions (open / closed eyes, smiling / not smiling) and facial
    details (glasses / no glasses). All the images were taken against a dark
    homogeneous background with the subjects in an upright, frontal position
    (with tolerance for some side movement).

**Data Set Characteristics:**

=================   =====================
Classes              

In [ ]:
# Step 3: Access the target labels of the dataset
# Hint: Use the `.target` attribute of the `olivetti` object to get the labels.
# These labels are integers from 0 to 39, identifying the person in each image.

In [ ]:
olivetti.target

array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  3,  3,  3,  3,
        3,  3,  3,  3,  3,  3,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  5,
        5,  5,  5,  5,  5,  5,  5,  5,  5,  6,  6,  6,  6,  6,  6,  6,  6,
        6,  6,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  8,  8,  8,  8,  8,
        8,  8,  8,  8,  8,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9, 10, 10,
       10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 11, 11, 11, 11, 11, 11, 11,
       11, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13, 13,
       13, 13, 13, 13, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 15, 15, 15,
       15, 15, 15, 15, 15, 15, 15, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16,
       17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 18, 18, 18, 18, 18, 18, 18,
       18, 18, 18, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 20, 20, 20, 20,
       20, 20, 20, 20, 20, 20, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 22,
       22, 22, 22, 22, 22

### What is Stratified Sampling?
Stratified sampling is a technique used when splitting a dataset into subsets (like training, validation, and test sets) to preserve the distribution of the target labels.

### Why do we use it?
In classification problems—especially with imbalanced or small datasets like Olivetti Faces—you want to ensure that each class (person) is represented equally in every subset of your data (training, validation, test).
Without it, some classes might appear more in one split than others, which can lead to biased model performance.

**This ensures** that the proportion of images for each of the 40 people is approximately the same in the training, validation, and test sets.

Example:
If person 0 appears 10 times in the dataset:

With stratified sampling, person 0 will appear 8 times in the 80% training+validation set and 2 times in the 20% test set.

Without it, you could randomly end up with, say, all of person 0’s images in just the training set—and none in validation or test—which is bad for generalization.

In [ ]:
# Step 5: Use stratified sampling to split the dataset manually
# You will use `StratifiedShuffleSplit` from `sklearn.model_selection` to ensure
# each class (person) is equally represented in all splits.


# First split: Create test set with 40 samples (1 image per person)
# Method used: StratifiedShuffleSplit(n_splits=1, test_size=40)
# Variable: strat_split defines the split object

# Use .split(X, y) to generate indices for train/validation and test sets
# Variables: train_valid_idx and test_idx

# Use the indices to extract the corresponding data and labels

#  Second split: Split train_valid into 80 training and 80 validation samples
# Method used: StratifiedShuffleSplit(n_splits=1, test_size=80)
# This ensures each person still has 2 images in both training and validation sets.

# Split training/validation data using the new stratified splitter



# Create the final training and validation sets
X_train =
y_train =
X_valid =
y_valid =

# Output the number of samples in each set

print("Training set size":,)
print("Validation set size":,)
print("Test set size":,)



In [ ]:
#  Step 5: Use stratified sampling to split the dataset manually
# You will use `StratifiedShuffleSplit` from `sklearn.model_selection` to ensure
# each class (person) is equally represented in all splits.

from sklearn.model_selection import StratifiedShuffleSplit

#  First split: Create test set with 40 samples (1 image per person)
# Method used: StratifiedShuffleSplit(n_splits=1, test_size=40)
# Variable: strat_split defines the split object
strat_split = StratifiedShuffleSplit(n_splits=1, test_size=40, random_state=42)

# Use .split(X, y) to generate indices for train/validation and test sets
# Variables: train_valid_idx and test_idx
train_valid_idx, test_idx = next(strat_split.split(olivetti.data, olivetti.target))

# Use the indices to extract the corresponding data and labels
X_train_valid = olivetti.data[train_valid_idx]
y_train_valid = olivetti.target[train_valid_idx]
X_test = olivetti.data[test_idx]
y_test = olivetti.target[test_idx]

#  Second split: Split train_valid into 80 training and 80 validation samples
# Method used: StratifiedShuffleSplit(n_splits=1, test_size=80)
# This ensures each person still has 2 images in both training and validation sets.
strat_split = StratifiedShuffleSplit(n_splits=1, test_size=80, random_state=43)

# Split training/validation data using the new stratified splitter
train_idx, valid_idx = next(strat_split.split(X_train_valid, y_train_valid))

# Create the final training and validation sets
X_train = X_train_valid[train_idx]
y_train = y_train_valid[train_idx]
X_valid = X_train_valid[valid_idx]
y_valid = y_train_valid[valid_idx]

# 🧾 Output the number of samples in each set
print("Training set size:", X_train.shape)
print("Validation set size:", X_valid.shape)
print("Test set size:", X_test.shape)


Training set size: (280, 4096)
Validation set size: (80, 4096)
Test set size: (40, 4096)


###  Dataset Split Summary

After applying **Stratified Sampling**, the Olivetti Faces dataset is divided into three sets:

| Dataset        | Number of Samples | Shape (samples × features) | Description |
|----------------|-------------------|-----------------------------|-------------|
| **Training**   | 280               | (280, 4096)                 | Used to train the model |
| **Validation** | 80                | (80, 4096)                  | Used to tune model parameters and prevent overfitting |
| **Test**       | 40                | (40, 4096)                  | Used to evaluate the final model performance |

- Each image is originally **64 × 64 pixels**, flattened into a vector of **4096** pixel values (features).
- There are a total of **400 images**, evenly representing **40 individuals (10 images per person)**.
- We use **stratified sampling** to maintain equal class distribution across all splits.


# PCA for data dimentionality

To speed things up, we'll reduce the data's dimensionality using PCA:

###  Dimensionality Reduction with PCA

To speed things up and simplify our analysis, we'll use **Principal Component Analysis (PCA)** to reduce the dataset's dimensionality.

- Each face image currently has **4096 features** (one for each pixel).
- This high dimensionality can be computationally expensive and may include redundant information.
- PCA helps us:
  -  **Compress** the data into fewer dimensions,
  -  **Capture the most important variance** in the dataset,
  -  **Improve clustering performance and speed**.

We’ll use **`sklearn.decomposition.PCA`** to reduce the data to a smaller number of principal components.


-  What Does `PCA(n_components=0.99)` Mean?

Instead of manually choosing the number of dimensions to reduce to, we ask PCA to:
> 💬 "Keep enough components to preserve **99% of the variance** (information) in the original data."
- Why Transform X_valid and X_test?
After training PCA on the training set (X_train), we apply the same transformation to:
_Validation set (X_valid)
_Test set (X_test)
This ensures that all datasets are projected into the same reduced feature space, preventing data leakage and keeping the evaluation fair.

In [ ]:
# Step 1: Import the PCA class from sklearn.decomposition
# Method: PCA

# Step 2: Create a PCA object to preserve 99% of the variance
# Variable to use: pca
# Constructor method: PCA(n_components=0.99)

# Step 3: Fit PCA to the training set and transform it
# Method: fit_transform
# Variable: X_train_pca

# Step 4: Use the trained PCA to transform the validation set
# Method: transform
# Variable: X_valid_pca

# Step 5: Use the same PCA transformation on the test set
# Method: transform
# Variable: X_test_pca

# Step 6: Display the number of principal components chosen to preserve 99% of the variance
# Attribute: n_components_


In [ ]:
# Step 1: Import the PCA class from sklearn.decomposition
# Method: PCA
from sklearn.decomposition import PCA

# Step 2: Create a PCA object to preserve 99% of the variance
# Variable to use: pca
# Constructor method: PCA(n_components=0.99)
pca = PCA(n_components=0.99)

# Step 3: Fit PCA to the training set and transform it
# Method: fit_transform
# Variable: X_train_pca
X_train_pca = pca.fit_transform(X_train)

# Step 4: Use the trained PCA to transform the validation set
# Method: transform
# Variable: X_valid_pca
X_valid_pca = pca.transform(X_valid)

# Step 5: Use the same PCA transformation on the test set
# Method: transform
# Variable: X_test_pca
X_test_pca = pca.transform(X_test)

# Step 6: Display the number of principal components chosen to preserve 99% of the variance
# Attribute: n_components_
pca.n_components_


np.int64(199)

# K-means clustering
Cluster the images using K-Means, and ensure that you have a good number of clusters

## Training the model

In [ ]:
#  Step: Perform KMeans Clustering on PCA-Transformed Data
# -----------------------------------------------------------
# In this step, we will apply KMeans clustering to the Olivetti dataset that has been transformed using PCA.
# We'll test different values for 'k' (number of clusters) ranging from 5 to 145, in steps of 5.
#
#  Method: sklearn.cluster.KMeans
#  Variables to use:
#     - X_train_pca: your PCA-transformed training set
#     - kmeans_per_k: a list to store the resulting KMeans models for each 'k'
#
#  Your task:
# Loop through a range of 'k' values, create a KMeans model for each, fit it to X_train_pca, and store the fitted model in the list.


 # Try cluster counts from 5 to 145 (step = 5)
  # List to store fitted KMeans models

In [ ]:
#  Step: Perform KMeans Clustering on PCA-Transformed Data
# -----------------------------------------------------------
# In this step, we will apply KMeans clustering to the Olivetti dataset that has been transformed using PCA.
# We'll test different values for 'k' (number of clusters) ranging from 5 to 145, in steps of 5.
#
#  Method: sklearn.cluster.KMeans
#  Variables to use:
#     - X_train_pca: your PCA-transformed training set
#     - kmeans_per_k: a list to store the resulting KMeans models for each 'k'
#
#  Your task:
# Loop through a range of 'k' values, create a KMeans model for each, fit it to X_train_pca, and store the fitted model in the list.

from sklearn.cluster import KMeans

k_range = range(5, 150, 5)  # Try cluster counts from 5 to 145 (step = 5)
kmeans_per_k = []  # List to store fitted KMeans models

for k in k_range:
    print(f"k = {k}")
    kmeans = KMeans(n_clusters=k, random_state=42).fit(X_train_pca)
    kmeans_per_k.append(kmeans)


## Evaluating the model and fine-tuning


In [ ]:

#  Step: Evaluate KMeans Clustering Using Silhouette Score
# ----------------------------------------------------------
# Now that we have fitted KMeans models for different cluster counts (k),
# we will evaluate each model using the Silhouette score, which measures
# how well-separated the clusters are.
#
#  Methods and functions to use:
#   - sklearn.metrics.silhouette_score(X, labels)
#   - np.argmax() to find the best silhouette score index
#   - matplotlib.pyplot for plotting
#
#  Variables:
#   - X_train_pca: PCA-transformed training data
#   - kmeans_per_k: list of fitted KMeans models
#   - k_range: list/range of k values tried
#
#  Your task:
# 1. Calculate silhouette scores for each model.
# 2. Find the best k with the highest silhouette score.
# 3. Plot the silhouette scores vs k and highlight the best k.





In [ ]:
from sklearn.metrics import silhouette_score
import numpy as np
import matplotlib.pyplot as plt

silhouette_scores = [silhouette_score(X_train_pca, model.labels_) for model in kmeans_per_k]

best_index = np.argmax(silhouette_scores)    # Index of highest silhouette score
best_k = k_range[best_index]                  # Corresponding best number of clusters
best_score = silhouette_scores[best_index]   # Best silhouette score value

plt.figure(figsize=(8, 3))
plt.plot(k_range, silhouette_scores, "bo-", label="Silhouette Score")
plt.xlabel("$k$", fontsize=14)
plt.ylabel("Silhouette Score", fontsize=14)
plt.plot(best_k, best_score, "rs", label=f"Best k = {best_k}")
plt.legend()
plt.title("Silhouette Scores for Different k Values")
plt.show()

In [ ]:
best_k

###  Why Is the Best Number of Clusters Much Higher Than 40?

You might expect the best number of clusters to be **40**, because there are exactly 40 different people in the dataset. However, the clustering algorithm suggests a much higher number (e.g., 120).

#### Why does this happen?

- **Variability within each person’s images:** The same person can look quite different in different photos. For example:
  - Wearing glasses in some images but not others
  - Different facial expressions or head poses
  - Lighting or shadows shifting the appearance
  - Slight shifts in alignment or cropping

- Because of these differences, the algorithm **splits images of the same person into multiple clusters** to better capture subtle variations.



###  What does this tell us?

- Real-world face images are complex and contain more variation than just the identity of the person.
- Clustering is capturing these **intra-person variations** along with inter-person differences.
- This highlights the importance of **dimensionality reduction and feature extraction** before clustering or classification tasks.



###  Next steps:

- You might want to try other clustering algorithms or tweak parameters.
- Or move towards supervised learning methods that explicitly learn to recognize identities despite variations.


### Understanding Inertia in KMeans Clustering

**Inertia** is a measure of how well the clustering has grouped the data points.

- It is the **sum of squared distances** between each data point and the centroid of the cluster it belongs to.
- A **lower inertia value** means that the points are closer to their cluster centers, indicating more compact and well-defined clusters.
- However, inertia always decreases as you increase the number of clusters (k), so it should be used together with other metrics like the silhouette score to find a good balance.

**What we are trying to do:**  
By plotting inertia against different numbers of clusters (`k`), we look for an "elbow point" where the inertia stops decreasing significantly. This helps us choose a number of clusters that balances compactness with simplicity, avoiding too many clusters that overfit the data.


In [ ]:
#  Step: Analyze KMeans Clustering Using Inertia (Within-Cluster Sum of Squares)
# ------------------------------------------------------------------------------
# Inertia measures how tightly grouped the points in each cluster are.
# Lower inertia means clusters are more compact.
#
#  Methods and attributes:
#   - model.inertia_: attribute from each fitted KMeans model
#   - matplotlib.pyplot for plotting
#
#  Variables:
#   - kmeans_per_k: list of fitted KMeans models for different k values
#   - k_range: range/list of k values tested
#   - best_index: index of the best model based on silhouette score (from previous step)
#
#  Your task:
# 1. Extract inertia values for each model in kmeans_per_k.
# 2. Identify the inertia value of the best model (best_index).
# 3. Plot inertia vs k and highlight the best model's inertia.



### Elbow Diagram

The **elbow diagram** is a plot of the **inertia** values for different numbers of clusters (`k`) in KMeans clustering.

- **Inertia** measures how tightly data points are grouped around their cluster centers (lower is better).
- As `k` increases, inertia typically decreases because more clusters mean smaller, tighter groups.
- The **elbow point** is where the decrease in inertia slows down significantly.
- This "elbow" helps us select an optimal number of clusters by balancing compactness and simplicity.

If the elbow is not clear, we may need to consider other metrics or domain knowledge to choose `k`.


In [ ]:

inertias = [model.inertia_ for model in kmeans_per_k]
best_inertia = inertias[best_index]

plt.figure(figsize=(8, 3.5))
plt.plot(k_range, inertias, "bo-", label="Inertia")
plt.xlabel("$k$", fontsize=14)
plt.ylabel("Inertia", fontsize=14)
plt.plot(best_k, best_inertia, "rs", label=f"Best k = {best_k}")
plt.legend()
plt.title("KMeans Inertia for Different Numbers of Clusters")
plt.show()

The optimal number of clusters is not clear on this inertia diagram, as there is no obvious elbow, so let's stick with k=120.

### Visualize the clusters:
**do you see similar faces in each cluster?**

In [ ]:
# Step 1: Define a function `plot_faces` that:
# - Takes the following arguments:
#     - `faces`: a NumPy array of flattened face images (shape: [n_samples, 4096])
#     - `labels`: corresponding labels or cluster IDs (array-like)
#     - `n_cols`: number of columns to display per row (default: 5)
# - Inside the function:
#     - Use the `reshape` method on `faces` to convert each flattened image to 64x64 pixels
#     - Calculate the number of rows needed for the grid layout
#     - Use matplotlib's `plt.figure()` to create the figure
#     - Loop over the images and labels with `enumerate()` and `zip()`
#       - For each face:
#         - Use `plt.subplot()` to create a subplot in the grid
#         - Display the image with `plt.imshow()`, set `cmap="gray"`
#         - Turn off axis ticks with `plt.axis("off")`
#         - Set the title with `plt.title()` to show the label
#     - Use `plt.show()` to display the full grid of images






# Step 2: Loop over unique cluster IDs from `best_model.labels_` using `np.unique()`
# - For each cluster ID:
#     - Create a boolean mask `in_cluster` where `best_model.labels_ == cluster_id`
#     - Select corresponding faces from `X_train` using this mask
#     - Select corresponding labels from `y_train` using the same mask
#     - Call the `plot_faces()` function to visualize the cluster faces with labels




In [ ]:
def plot_faces(faces, labels, n_cols=5):
    faces = faces.reshape(-1, 64, 64)
    n_rows = (len(faces) - 1) // n_cols + 1
    plt.figure(figsize=(n_cols, n_rows * 1.1))
    for index, (face, label) in enumerate(zip(faces, labels)):
        plt.subplot(n_rows, n_cols, index + 1)
        plt.imshow(face, cmap="gray")
        plt.axis("off")
        plt.title(label)
    plt.show()

for cluster_id in np.unique(best_model.labels_):
    print("Cluster", cluster_id)
    in_cluster = best_model.labels_==cluster_id
    faces = X_train[in_cluster]
    labels = y_train[in_cluster]
    plot_faces(faces, labels)

About 2 out of 3 clusters are useful: that is, they contain at least 2 pictures, all of the same person. However, the rest of the clusters have either one or more intruders, or they have just a single picture.

Clustering images this way may be too imprecise to be directly useful when training a model (as we will see below), but it can be tremendously useful when labeling images in a new dataset: it will usually make labelling much faster.

## 11. Using Clustering as Preprocessing for Classification

Continuing with the Olivetti faces dataset, train a classifier to predict which person is represented in each picture, and evaluate it on the validation set.*

In [ ]:
# Step 1: Import the RandomForestClassifier from sklearn.ensemble.

# Step 2: Create a RandomForestClassifier instance named `clf` with:
#         - `n_estimators=150` (number of trees)
#         - `random_state=42` for reproducibility.

# Step 3: Train the classifier using the `.fit()` method on the PCA-transformed training data:
#         - Features: `X_train_pca`
#         - Labels: `y_train`

# Step 4: Evaluate the trained model using the `.score()` method on the PCA-transformed validation set:
#         - Features: `X_valid_pca`
#         - Labels: `y_valid`
#         - This returns the accuracy score.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(n_estimators=150, random_state=42)
clf.fit(X_train_pca, y_train)
clf.score(X_valid_pca, y_valid)

Next, use K-Means as a dimensionality reduction tool, and train a classifier on the reduced set.*

In [ ]:
# Step 1: Use the best KMeans model stored in `best_model` to transform your PCA-reduced data.
#         This applies clustering-based dimensionality reduction.
#         - Transform the training set `X_train_pca` into `X_train_reduced`
#         - Transform the validation set `X_valid_pca` into `X_valid_reduced`
#         - Transform the test set `X_test_pca` into `X_test_reduced`
#         (Hint: Use the `.transform()` method of the KMeans model.)

# Step 2: Import `RandomForestClassifier` from `sklearn.ensemble`.

# Step 3: Instantiate the classifier `clf` with 150 estimators and a fixed `random_state=42`.

# Step 4: Fit the classifier on the reduced training data `X_train_reduced` with labels `y_train` using `.fit()`.

# Step 5: Evaluate the model accuracy on the reduced validation data `X_valid_reduced` and labels `y_valid` using `.score()`.


In [ ]:
X_train_reduced = best_model.transform(X_train_pca)
X_valid_reduced = best_model.transform(X_valid_pca)
X_test_reduced = best_model.transform(X_test_pca)

clf = RandomForestClassifier(n_estimators=150, random_state=42)
clf.fit(X_train_reduced, y_train)

clf.score(X_valid_reduced, y_valid)

 That's not better at all! Let's see if tuning the number of clusters helps.

We could use a `GridSearchCV` like we did earlier in this notebook, but since we already have a validation set, we don't need K-fold cross-validation, and we're only exploring a single hyperparameter, so it's simpler to just run a loop manually:

In [ ]:
# Step 1: Import Pipeline from sklearn.pipeline.

# Step 2: For each number of clusters in the list `k_range`, do the following:
#   - Create a pipeline named `pipeline` that includes:
#       - A KMeans clustering step named "kmeans" with `n_clusters` clusters.
#       - A RandomForestClassifier step named "forest_clf" with 150 trees.
#   - Fit the pipeline on the PCA-reduced training data (`X_train_pca`) and training labels (`y_train`) using `.fit()`.
#   - Print the current number of clusters (`n_clusters`) and the validation accuracy obtained by scoring on `X_valid_pca` and `y_valid` using `.score()`.


In [ ]:
from sklearn.pipeline import Pipeline

for n_clusters in k_range:
    pipeline = Pipeline([
        ("kmeans", KMeans(n_clusters=n_clusters, random_state=42)),
        ("forest_clf", RandomForestClassifier(n_estimators=150, random_state=42))
    ])
    pipeline.fit(X_train_pca, y_train)
    print(n_clusters, pipeline.score(X_valid_pca, y_valid))

Oh well, even by tuning the number of clusters, we never get beyond 80% accuracy. Looks like the distances to the cluster centroids are not as informative as the original images.

### Combining PCA Features with Cluster Membership Features

In our previous steps, we reduced the dimensionality of the face images using **PCA**, which gives us a compact and important representation of each face. Then, we applied **KMeans clustering** to group similar faces into clusters.

But instead of using only the PCA features for classification, we can improve our model by also including information about the cluster each face belongs to. Here’s the idea:

1. **PCA features** represent the continuous characteristics of each face image in fewer dimensions.
2. **KMeans clustering** groups faces into clusters based on similarity.
3. For each face, we find which cluster it belongs to (its cluster label).
4. We convert these cluster labels into **one-hot encoded vectors** (e.g., if there are 5 clusters, cluster 2 becomes `[0, 1, 0, 0, 0]`).
5. We **append** these one-hot cluster vectors to the PCA features, creating an extended feature set that combines both types of information.
6. Finally, we train our classifier using this combined feature set.

---

### Why does this help?

- PCA captures *detailed, continuous* features about each face.
- Cluster membership adds *categorical* information about which group of similar faces an image belongs to.
- Combining these features provides the classifier with richer information, potentially improving prediction accuracy.

Think of it like giving the model both a detailed description and a group label — it can use both to make better decisions!


In [ ]:
# Step: Combine the original PCA features with the cluster membership features (one-hot encoded)
# Hint: Use numpy's c_ method to concatenate arrays column-wise.
# Variables:
# - X_train_pca: PCA features for the training set
# - X_train_reduced: Cluster membership features (one-hot encoded) for the training set
# - Similarly for validation and test sets

In [ ]:
X_train_extended = np.c_[X_train_pca, X_train_reduced]
X_valid_extended = np.c_[X_valid_pca, X_valid_reduced]
X_test_extended = np.c_[X_test_pca, X_test_reduced]

In [ ]:
# Step: Train a RandomForestClassifier on the extended feature set
# Hint: Use RandomForestClassifier from sklearn.ensemble with 150 trees and random_state=42 for reproducibility
# Variables:
# - X_train_extended: Training features combining PCA + cluster features
# - y_train: Training labels
# - X_valid_extended: Validation features combining PCA + cluster features
# - y_valid: Validation labels

In [ ]:
clf = RandomForestClassifier(n_estimators=150, random_state=42)
clf.fit(X_train_extended, y_train)
clf.score(X_valid_extended, y_valid)

That's a bit better, but still worse than without the cluster features. The clusters are not useful to directly train a classifier in this case (but they can still help when labelling new training instances).

## 12. A Gaussian Mixture Model for the Olivetti Faces Dataset

Train a Gaussian mixture model on the Olivetti faces dataset. To speed up the algorithm, you should probably reduce the dataset's dimensionality (e.g., use PCA, preserving 99% of the variance).

In [ ]:
from sklearn.mixture import GaussianMixture

gm = GaussianMixture(n_components=40, random_state=42)
y_pred = gm.fit_predict(X_train_pca)

Use the model to generate some new faces (using the sample() method), and visualize them (if you used PCA, you will need to use its inverse_transform() method).

In [ ]:
n_gen_faces = 20
gen_faces_reduced, y_gen_faces = gm.sample(n_samples=n_gen_faces)
gen_faces = pca.inverse_transform(gen_faces_reduced)

In [ ]:
plot_faces(gen_faces, y_gen_faces)

Try to modify some images (e.g., rotate, flip, darken) and see if the model can detect the anomalies (i.e., compare the output of the score_samples() method for normal images and for anomalies).

In [ ]:
n_rotated = 4
rotated = np.transpose(X_train[:n_rotated].reshape(-1, 64, 64), axes=[0, 2, 1])
rotated = rotated.reshape(-1, 64*64)
y_rotated = y_train[:n_rotated]

n_flipped = 3
flipped = X_train[:n_flipped].reshape(-1, 64, 64)[:, ::-1]
flipped = flipped.reshape(-1, 64*64)
y_flipped = y_train[:n_flipped]

n_darkened = 3
darkened = X_train[:n_darkened].copy()
darkened[:, 1:-1] *= 0.3
y_darkened = y_train[:n_darkened]

X_bad_faces = np.r_[rotated, flipped, darkened]
y_bad = np.concatenate([y_rotated, y_flipped, y_darkened])

plot_faces(X_bad_faces, y_bad)

In [ ]:
X_bad_faces_pca = pca.transform(X_bad_faces)

In [ ]:
gm.score_samples(X_bad_faces_pca)

The bad faces are all considered highly unlikely by the Gaussian Mixture model. Compare this to the scores of some training instances:

In [ ]:
gm.score_samples(X_train_pca[:10])

 # Using Dimensionality Reduction Techniques for Anomaly Detection


 In this section, we use Principal Component Analysis (PCA) to reconstruct images and measure how well the model can represent each face. By comparing the original images with their PCA reconstructions, we can identify anomalies or unusual images — those that PCA struggles to represent accurately.

The reconstruction error (mean squared error between the original and reconstructed image) is a key metric:

Low error indicates that the image fits well within the learned patterns (normal faces).

High error suggests the image may be noisy, corrupted, or an outlier (anomalous face).

This technique is useful in real-world applications such as:

Detecting corrupted or poor-quality images in biometric security systems.

Filtering out invalid or unusual data before training face recognition models.

Some dimensionality reduction techniques can also be used for anomaly detection. For example, take the Olivetti faces dataset and reduce it with PCA, preserving 99% of the variance. Then compute the reconstruction error for each image. Next, take some of the modified images you built in the previous exercise, and look at their reconstruction error: notice how much larger the reconstruction error is. If you plot a reconstructed image, you will see why: it tries to reconstruct a normal face.

We already reduced the dataset using PCA earlier:

In [ ]:
X_train_pca

In [ ]:
# Step: Define a function to compute reconstruction errors using PCA
# Hint:
# - Use pca.transform() to reduce dimensionality of input data X
# - Use pca.inverse_transform() to reconstruct the original data from reduced data
# - Compute mean squared error (MSE) between reconstructed and original data for each sample
# Variables:
# - pca: fitted PCA object
# - X: input data to compute reconstruction error on
# Returns:
# - mse: array of reconstruction errors for each sample

In [ ]:
def reconstruction_errors(pca, X):
    X_pca = pca.transform(X)
    X_reconstructed = pca.inverse_transform(X_pca)
    mse = np.square(X_reconstructed - X).mean(axis=-1)
    return mse

In [ ]:
# Step: Calculate the average reconstruction error on the training set
# Hint:
# - Use the reconstruction_errors() function you defined earlier
# - Pass the fitted PCA object 'pca' and training data 'X_train' as arguments
# - Use .mean() to get the average reconstruction error across all training samples
# Variables:
# - pca: the fitted PCA object
# - X_train: training dataset (original features)
# - Output: average reconstruction error (a single float number)

In [ ]:
reconstruction_errors(pca, X_train).mean()

In [ ]:
# Step: Calculate the average reconstruction error for a set of "bad" or noisy face images
# Hint:
# - Use the previously defined function reconstruction_errors()
# - Pass the fitted PCA object 'pca' and the dataset 'X_bad_faces' as arguments
# - Calculate the mean error to summarize the overall reconstruction quality
# Variables:
# - pca: the fitted PCA object
# - X_bad_faces: dataset containing distorted or noisy faces to test the PCA reconstruction
# - Output: average reconstruction error for these "bad" faces (a single float number)

In [ ]:
reconstruction_errors(pca, X_bad_faces).mean()

In [ ]:
# Step: Visualize the "bad" or noisy face images using the provided plot_faces() function
# Hint:
# - Use the function plot_faces()
# - Pass the dataset 'X_bad_faces' and the corresponding labels 'y_bad'
# - This will help you visually inspect how the noisy faces look compared to normal ones
# Variables:
# - X_bad_faces: the set of noisy or distorted face images
# - y_bad: the corresponding labels or IDs for these images

In [ ]:
plot_faces(X_bad_faces, y_bad)

In [ ]:
# Step: Reconstruct the "bad" faces from their PCA representation
# Hint:
# - Use the PCA method `inverse_transform()` to convert reduced data back to original space
# - Store the reconstructed faces in the variable `X_bad_faces_reconstructed`
# Variables:
# - X_bad_faces_pca: PCA-transformed representation of the noisy faces
# - X_bad_faces_reconstructed: reconstructed faces after inverse PCA transform


# Step: Visualize the reconstructed "bad" faces to compare with original noisy images
# Hint:
# - Use the `plot_faces()` function again
# - Pass `X_bad_faces_reconstructed` and `y_bad` as inputs



PCA Reconstruction of Faces
In this step, we take the PCA-reduced representation of the “bad” or noisy face images (X_bad_faces_pca) and reconstruct them back into the original high-dimensional space using the PCA method inverse_transform(). This process attempts to recover the original images from their compressed form.

Why do we do this?
PCA reduces the dimensionality of the data by keeping only the most important features. When we reconstruct the images, we can see how much information was retained or lost after this compression.

What does the visualization show?
By plotting the reconstructed images (X_bad_faces_reconstructed), we can visually compare them with the original noisy images (X_bad_faces). This helps us understand the quality of the dimensionality reduction and how well PCA captures the essential features of the faces despite noise or distortions.

